wire-ios-data-model/Source/Model/Conversation/ZMConversation+Participants.swift

```diff
@@ -139,18 +140,42 @@ extension ZMConversation {
     }
 
     // MARK: - Participant actions
+    public func addParticipants(_ participants: [UserType],
+                                completion: @escaping AddParticipantAction.ResultHandler) {
 
-    public func addParticipants(
-        _ participants: [UserType],
-        completion: @escaping AddParticipantAction.ResultHandler
-    ) {
         guard let context = managedObjectContext else {
             completion(.failure(.unknown))
             return
         }
-
         let users = participants.materialize(in: context)
 
+        internalAddParticipants(users) { (result) in
+            switch result {
+            case .success:
+                return completion(result)
+            case .failure(let error):
+                switch error {
+                case .unreachableUsers(let unreachableUsers):
+                    let reachableUsers = Set(users).subtracting(Set(unreachableUsers))
+                    self.appendFailedToAddUsersSystemMessage(users: Set(unreachableUsers),
+                                                             sender: self.creator,
+                                                             at: self.lastServerTimeStamp ?? Date())
+
+                    self.internalAddParticipants(Array(reachableUsers), completion: completion)
+                default:
+                    return completion(result)
+                }
+            }
+        }
+    }
+
+    private func internalAddParticipants(_ users: [ZMUser],
```

Comment:
Change this to `Set<ZMUser>` and you won't have to do some casting when calling the method.

Reply:
We need to have an array of users inside `internalAddParticipants(...)`. We create an AddParticipantAction action and it uses an array as a parameter for the initialiser.